client/components/notice/index.tsx

```diff
@@ -0,0 +1,124 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, FunctionComponent, useRef } from 'react';
+import classnames from 'classnames';
+import { useTranslate } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import { TimerHandle } from 'calypso/types';
+import Gridicon from 'calypso/components/gridicon';
+import { NoticeStatus, NoticeText } from './types';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+/**
+ * Module constants
+ */
+const GRIDICONS_WITH_DROP = [
+	'add',
+	'cross-circle',
+	'ellipsis-circle',
+	'help',
+	'info',
+	'notice',
+	'pause',
+	'play',
+	'spam',
+];
+
+interface Props {
+	className?: string;
+	duration?: number;
+	onDismissClick: () => void;
+	children: React.ReactNode;
+	status?: NoticeStatus;
+	icon?: string | null;
+	isCompact?: boolean;
+	isLoading?: boolean;
+	showDismiss?: boolean;
+	text?: NoticeText;
+}
+
+const getIcon = ( status: NoticeStatus ): string => {
+	switch ( status ) {
+		case 'is-info':
+			return 'info';
+		case 'is-success':
+			return 'checkmark';
+		case 'is-error':
+			return 'notice';
+		case 'is-warning':
+			return 'notice';
+		case 'is-plain':
+			return 'info';
+		default:
+			return 'info';
+	}
+};
+
+const Notice: FunctionComponent< Props > = ( {
+	className,
+	duration = 0,
+	onDismissClick,
+	children,
+	status,
+	icon,
+	isCompact = false,
+	isLoading = false,
+	showDismiss = ! isCompact, // by default, show on normal notices, don't show on compact ones
+	text,
+} ) => {
+	const translate = useTranslate();
+	const dismissTimeout = useRef< TimerHandle | null >( null );
+
+	useEffect( () => {
+		if ( duration > 0 ) {
+			dismissTimeout.current = setTimeout( onDismissClick, duration );
+		}
+
+		return () => {
+			if ( dismissTimeout.current ) {
+				clearTimeout( dismissTimeout.current );
+			}
+		};
+	}, [ duration, onDismissClick ] );
```

Comment:
You probably don't want to reset the timeout when `onDismissClick` callback changes. Better to store it in a ref:
```js
useEffect( () => {
  dismissCallback.current = onDismissClick;
}, [ onDismissClick ] );

useEffect( () => {
  let id = setTimeout( () => dismissCallback.current(), duration );
  return () => clearTimeout( id );
}, [ duration ] );
```
Dan Abramov has a nice article about this where he explains everything:
https://overreacted.io/making-setinterval-declarative-with-react-hooks/